compiler/rustc_infer/src/infer/error_reporting/nice_region_error/different_lifetimes.rs

```diff
@@ -102,43 +102,73 @@ impl<'a, 'tcx> NiceRegionError<'a, 'tcx> {
-                        "...but data with one lifetime flows into the other here".to_owned(),
+                        ty_sub.span,
+                        ret_span,
+                        format!(
+                            "this parameter and the {} are declared with different lifetimes...",
+                            return_type
+                        ),
+                        format!("...but data{} is {} here", span_label_var1, action),
+                        sup_future,
                     )
-                } else {
+                }
+                (_, Some(ret_span)) => {
+                    let sub_future = self.future_return_type(scope_def_id_sub);
+                    let (return_type, action) = if let Some(_) = sub_future {
+                        ("returned future", "held across an await point")
+                    } else {
+                        ("return type", "returned")
+                    };
+
                     (
-                        "these two types are declared with different lifetimes...".to_owned(),
-                        format!("...but data{} flows{} here", span_label_var1, span_label_var2),
+                        ty_sup.span,
+                        ret_span,
+                        format!(
+                            "this parameter and the {} are declared with different lifetimes...",
+                            return_type
+                        ),
+                        format!("...but data{} is {} here", span_label_var1, action),
+                        sub_future,
                     )
-                };
-                (ty_sup.span, ty_sub.span, main_label_1, span_label_1)
-            }
-
-            (Some(ret_span), _) => (
-                ty_sub.span,
-                ret_span,
-                "this parameter and the return type are declared with different lifetimes..."
-                    .to_owned(),
-                format!("...but data{} is returned here", span_label_var1),
-            ),
-            (_, Some(ret_span)) => (
-                ty_sup.span,
-                ret_span,
-                "this parameter and the return type are declared with different lifetimes..."
-                    .to_owned(),
-                format!("...but data{} is returned here", span_label_var1),
-            ),
-        };
+                }
+            };
 
-        struct_span_err!(self.tcx().sess, span, E0623, "lifetime mismatch")
-            .span_label(span_1, main_label)
-            .span_label(span_2, String::new())
-            .span_label(span, span_label)
-            .emit();
+        let mut e = struct_span_err!(self.tcx().sess, span, E0623, "lifetime mismatch");
+
+        e.span_label(span_1, main_label);
+        e.span_label(span_2, String::new());
+        e.span_label(span, span_label);
+
+        if let Some(s) = future_span {
+            e.span_note(s, "`async fn`s return an `impl Future<Output={return type}`");
```

Comment:
```suggestion
            e.span_note(s, "`async fn`s return an `impl Future<Output = {return type}>`");
```

Also, could we get the return type itself and use it as part of the output? Like

```
`async fn`s return an `impl Future<Output = i32>`
```

That way it ties the type a bit closer to what the user actually wrote.

Reply:
so basically get the snippet from the span, but precheck if the type is `()` and hardcode it to `()` for the no `->` case right?

Reply:
for posterity, is there a standard way to make a human-readable (or default to something like `{unnamed_type}` for a `Ty`?

Reply:
`Ty<'_>` implements `Display`.

Reply:
I think this might be more readable (with the return type inlined too as estebank suggested):

```suggestion
            e.span_note(s, "`async fn`s implicitly return `impl Future<Output = {return type}>`");
```

Reply:
@estebank is that rustc_hir::Ty or rustc_middle::Ty? I can't find a Display impl on either, is there a helper somewhere that does something similar?

Reply:
I believe it's on [`rustc_middle::Ty`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/type.Ty.html#impl-Display).

Reply:
actually the snippet ended up working

Reply:
It might be easier and less bug-prone to use `Display` – I'm not super experienced with this part of the compiler, but is that right @estebank?